src/connector.rs

```diff
@@ -637,3 +543,145 @@ fn tls_wait(mut tls: MutexGuard<'_, TlsStream>) -> io::Result<()> {
+pub struct ServerAddress(Url);
+
+/// Capability to convert into a list of NATS server addresses.
+///
+/// There are several implementations ensuring the easy passing of one or more server addresses to
+/// functions like [`crate::connect()`].
+pub trait IntoServerList {
+    fn into_server_list(self) -> io::Result<Vec<ServerAddress>>;
+}
+
+impl FromStr for ServerAddress {
+    type Err = io::Error;
+
+    /// Parse an address of a NATS server.
+    ///
+    /// If not stated explicitly the `nats://` schema and port `4222` is assumed.
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let url: Url = if !input.contains("://") {
+            format!("nats://{}", input).parse()
+        } else {
+            input.parse()
+        }
+        .map_err(|e| {
+            Error::new(
+                ErrorKind::InvalidInput,
+                format!("NATS server URL is invalid: {}", e),
+            )
+        })?;
+
+        Self::from_url(url)
+    }
+}
+
+impl ServerAddress {
+    /// Check if the URL is a valid NATS server address.
+    pub fn from_url(url: Url) -> io::Result<Self> {
+        if url.scheme() != "nats" && url.scheme() != "tls" {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                format!("invalid scheme for NATS server URL: {}", url.scheme()),
+            ));
+        }
+
+        Ok(Self(url))
+    }
+
+    /// Turn the server address into a standard URL.
+    pub fn into_inner(self) -> Url {
+        self.0
+    }
+
+    /// Returns if tls is required by the client for this server.
+    pub fn tls_required(&self) -> bool {
+        self.0.scheme() == "tls"
+    }
+
+    /// Returns if the server url had embedded username and password.
+    pub fn has_user_pass(&self) -> bool {
+        self.0.username() != ""
+    }
+
+    /// Returns the host.
+    pub fn host(&self) -> &str {
+        self.0.host_str().unwrap()
+    }
+
+    /// Returns the port.
+    pub fn port(&self) -> u16 {
+        self.0.port().unwrap_or(4222)
+    }
+
+    /// Returns the optional username in the url.
+    pub fn username(&self) -> Option<SecureString> {
+        let user = self.0.username();
+        if user.is_empty() {
+            None
+        } else {
+            Some(SecureString::from(user.to_string()))
+        }
+    }
+
+    /// Returns the optional password in the url.
+    pub fn password(&self) -> Option<SecureString> {
+        self.0
+            .password()
+            .map(|password| SecureString::from(password.to_string()))
+    }
+
+    /// Return the sockets from resolving the server address.
+    ///
+    /// # Fault injection
+    ///
+    /// If compiled with the `"fault_injection"` feature this method might fail artificially.
+    pub fn socket_addrs(&self) -> io::Result<impl Iterator<Item = SocketAddr>> {
+        inject_io_failure().and_then(|_| (self.host(), self.port()).to_socket_addrs())
+    }
+}
+
+impl<'s> IntoServerList for &'s str {
```

Comment:
Would be really nice to have tests for all those impl's.

Reply:
The implementations are tested in the doctests of `nats::connect()`. Is this enough?

Reply:
yes. thanks!